### ISSUE TITLE
Test suite with tests using dependency and priority has wrong behavior

### ISSUE THREAD

**Issue by mhereu** [cbeust/testng#793]
Hi, I have a test class like the follow:
```java
package com.test.data;

import org.testng.annotations.Test;

/**
 * Created by martin on 2015-09-03.
 */
public class testBug {

    @Test(groups = {"bug"})
    public void testA() {
        System.out.println(">>> 1 <<<");
    }

    @Test(groups = {"bug"}, dependsOnMethods = {"testA"})
    public void testB() {
        System.out.println(">>> 2 <<<");
    }

    @Test(groups = {"bug"}, priority = 1)
    public void testC() {
        System.out.println(">>> 3 <<<");
    }

    @Test(groups = {"bug"}, priority = 1)
    public void testD() {
        System.out.println(">>> 4 <<<");
    }

    @Test(groups = {"bug"}, priority = 2)
    public void testE() {
        System.out.println(">>> 5 <<<");
    }
}
```

The test default priority is 0 so I was expecting as output the following result:
```
>>> 1 <<<
>>> 2 <<<
>>> 3 <<<
>>> 4 <<<
>>> 5 <<<
```
but I got:
```
>>> 1 <<<
>>> 3 <<<
>>> 4 <<<
>>> 5 <<<
>>> 2 <<<
```
The test was executed using Maven 3.3.3, sureFire plugin 2.18.1, TestNG 6.9.6 and Java jdk1.8.0_51. Suite xml like:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="Test Suite" verbose="1" parallel="classes" thread-count="100">
  <test name="Regression">
    <packages>
      <package name="com.test.data"/>
    </packages>
    </test>
</suite>
```

*Acceptance criteria:* TestA and TestB has priority 0 so they should be executed before the one with higher priority.

**Comment by prvn133t** [comment]
This only happens when `dependsOnMethods` is used. When a `@test` method has `dependsOnMethods` it neglects the `priority`

**Comment by juherr** [comment]
Potential duplicate of #474

**Comment by mhereu** [comment]
@prvn133t if the current behavior is the expected, the code is working fine, but is not in the common sense way :)
In the Priority documentation said that all test has priority=0 by defaults, so I assumed that when I designed the test execution order, because of this issue I removed the dependsOnMethods and use Priority in all the tests.
@juherr I also tried that way and I had the same problem, the execution order has the not expected order. I think we should get a compilation error or something when we have dependsOnMethods and Priority in the same test.

**Comment by juherr** [comment]
@cbeust The issue is located into `DynamicGraph#getFreeNodes` where `testA`, `testC`, `testD` and `testE` are free nodes at the first call and run before `testB` which is blocked by `testA` and will be available at the second call.

In fact, `m_nodeComparator` (where priority is used) is only used to sort free nodes but not to filter them.

What should `getFreeNodes` do?
May the behavior change if tests are run in parallel or not?

**Comment by cbeust** [comment]
Yes, this looks very much like the bug I fixed in DynamicGraph for Kobalt,
which I didn't backport to TestNG. FYI, here is the current implementation:

https://github.com/cbeust/kobalt/blob/8acd8b8fcee4d763dd2552b154cf9c6c0097962e/modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/internal/DynamicGraph.kt#L41


-- 
Cédric
